aip/client-libraries/4210.md

```diff
@@ -31,25 +35,95 @@ expectation.
+  splines=ortho;
+  nodesep=0.3;
+  center=true;
+
+  proto [ label="API Protobuf\nDescriptors" shape=rectancle fillcolor=aliceblue ];
+
+  subgraph cluster_code_generators {
+    rank = same;
+    style = filled;
+    fillcolor = lightgrey;
+    node [ shape=oval ];
+
+    protobuf [ label="protobuf\ngenerator" fillcolor=deepskyblue3 ];
+    grpc [ label="gRPC\ngenerator" fillcolor=gold3 ];
+    gapic [ label="GAPIC\ngenerator" fillcolor=darkseagreen ];
+
+  }
+
+  proto -> protobuf;
+  proto -> grpc;
+  proto -> gapic;
+
+  subgraph cluster_generated_code {
+    rank = same;
+    style = filled;
+    fillcolor = lightgrey;
+    node [ shape=rectangle ];
+
+
+    protobuf_output [ label="Message & Enum\nCode" fillcolor=deepskyblue3 ];
+    grpc_output [ label="Server & Client\nStubs" fillcolor=gold3 ];
+    gapic_output [ label="Google API\nClient" fillcolor=darkseagreen ];
+
+  }
+  
+  protobuf -> protobuf_output;
+  grpc -> grpc_output;
+  gapic -> gapic_output;
+
+  assembly [ label="Package\nassembly" shape=oval fillcolor=aliceblue ];
+  
+  protobuf_output -> assembly
+  grpc_output -> assembly
+  gapic_output -> assembly
+
+  assembled_package [ label="Package of\ngenerated code" fillcolor=aliceblue ];
+
+  assembly -> assembled_package
+}
+```
+
+The following sections focus on the "GAPIC generator" in the above diagram.
+
 ### Protobuf plugins
 
-Code generators **must** be implemented as plugins to `protoc`, the protocol
-buffer compiler. The [plugin system][0] allows plugins to be written in any
-language, and plugins **should** ordinarily be written in the language being
-targeted, in order to take advantage of in-language tooling, and to ensure that
-experts in the target environment are able to meaningfully contribute.
+The protobuf compiler, `protoc`, supports a [plugin system][0] for code
+generation. The plugin system allows plugins to be written _in_ and _for_ any
+language.
+
+Code generators **must** be implemented as a `protoc` plugin. The following
+rules apply to the implementation of a client library generator as a `protoc`
+plugin:
 
+- The plugin  **should** be written in the language being targeted for
+  generation.
 - `protoc` expects plugins to be an executable in `$PATH`, and named
   `protoc-gen-{plugin_name}`, corresponding to the `--{plugin_name}_out` option
-  sent to the `protoc` executable.
-  - For a plugin creating client libraries for a specific language, the option
-    name **should** follow the convention `--{lang}_gapic_out` (meaning the
-    corresponding plugin executable is named `protoc-gen-{lang}_gapic`).
-- Plugins **must** accept a serialized `CodeGeneratorRequest` object (defined
+  sent to the `protoc` executable. As such:
+  - the plugin executable **should** be named `protoc-gen-{lang}_gapic`
+  - the plugin option  **should** follow the convention `--{lang}_gapic_out`
+- The plugin **must** accept a serialized `CodeGeneratorRequest` object (defined
```

Comment:
I'm not sure whether this bullet point is really worth including - we're really just saying "it must be a normal protoc plugin". I'd expect the rules in the AIP to be additional requirements. This bullet point and the next one could be in an explanatory section about what protoc does, but it feels a bit odd to have it here as rules... it's a bit like saying "For compiled languages, the generator must compile without errors."

Reply:
For sure, good point. I'll remove them.